networkx/lazy_imports.py

```diff
@@ -0,0 +1,179 @@
+
+      __getattr__, __lazy_dir__, __all__ = install_lazy(
+          __name__,
+          ['mysubmodule', 'anothersubmodule'],
+          {'foo': 'someattr'}
+      )
+
+    This functionality requires Python 3.7 or higher.
+
+    Parameters
+    ----------
+
+    module_name : str
+        Typically use __name__.
+    submodules : set
+        List of submodules to lazily import.
+    submod_attrs : dict
+        Dictionary of submodule -> list of attributes / functions.
+        These attributes are created the first time the submodule
+        is used.
+
+    Returns
+    -------
+
+    __getattr__, __dir__, __all__
+
+    """
+    if submod_attrs is None:
+        submod_attrs = {}
+
+    if submodules is None:
+        submodules = set()
+    else:
+        submodules = set(submodules)
+
+    attr_to_modules = {
+        attr: mod for mod, attrs in submod_attrs.items() for attr in attrs
+    }
+
+    __all__ = list(submodules | attr_to_modules.keys())
+
+    def __getattr__(name):
+        if name in submodules:
+            return importlib.import_module(f"{module_name}.{name}")
+        elif name in attr_to_modules:
+            submod = importlib.import_module(f"{module_name}.{attr_to_modules[name]}")
+            return getattr(submod, name)
+        else:
+            raise AttributeError(f"No {module_name} attribute {name}")
+
+    def __dir__():
+        return __all__
+
+    if os.environ.get("EAGER_IMPORT", ""):
+        for attr in set(attr_to_modules.keys()) | submodules:
+            __getattr__(attr)
+
+    return __getattr__, __dir__, list(__all__)
+
+
+class DelayedImportErrorLoader(importlib.util.LazyLoader):
+    def exec_module(self, module):
+        """Make the module load lazily."""
+        module.__spec__.loader = self.loader
+        module.__loader__ = self.loader
+        # Don't need to worry about deep-copying as trying to set an attribute
+        # on an object would have triggered the load,
+        # e.g. ``module.__spec__.loader = None`` would trigger a load from
+        # trying to access module.__spec__.
+        loader_state = {}
+        loader_state["__dict__"] = module.__dict__.copy()
+        loader_state["__class__"] = module.__class__
+        module.__spec__.loader_state = loader_state
+        module.__class__ = DelayedImportErrorModule
+
+
+class DelayedImportErrorModule(types.ModuleType):
+    """A subclass of ModuleType which trigger an ImportError upon attribute access
+
+    Based on importlib.util._LazyModule
+    """
+
+    def __getattribute__(self, attr):
+        spec = super().__getattribute__("__spec__")
+        raise ModuleNotFoundError(
+            f"Delayed Report: module named '{spec.name}' not found.\n"
+            "Reporting was Lazy -- delayed until module attributes accessed.\n"
+            f"Most likely, {spec.name} is not installed"
+        )
+
+
+def lazy_importorskip(modname, minversion=None, reason=None):
```

Comment:
```suggestion
def lazy_importerskip(modname, minversion=None, reason=None):
```

Reply:
It's unclear why pytest made this spelling mistake in their code!  I guess we can just fix ours.

Reply:
The spelling of `importorskip` bothered me when I first encountered pytest too.
But think of it as:  `import or skip`    It should probably be:   `import_or_skip`.
I think we should keep it the same as the word that pytest uses.
Though we could suggest the underscores to pytest.

Reply:
🤯 I would never have guessed the *or*! Yes, the underscores would be most helpful.